### PR TITLE
Handle CRDs created after Istio starts

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/30834

This is especially useful as we add support for more third party CRDs
(gateway-api and MCS). Currently, using these resources requires a
restart of Istiod. In normal cases, this is not ideal. In worst case
(external Istiod) this is impossible.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: tests/integration/pilot/main_test.go [hunk 1]
```diff
@@ -52,16 +51,6 @@ func supportsCRDv1(t resource.Context) bool {
 func TestMain(m *testing.M) {
 	framework.
 		NewSuite(m).
-		Setup(func(t resource.Context) (err error) {
```
Comment (howardjohn) [inline tests/integration/pilot/main_test.go:55]: This effectively tests this change - now we create the CRDs after install

File: pkg/config/schema/metadata.yaml [hunk 2]
```diff
@@ -256,7 +256,7 @@ snapshots:
 resources:
   # Kubernetes specific configuration.
   - kind: "CustomResourceDefinition"
-    plural: "CustomResourceDefinitions"
+    plural: "customresourcedefinitions"
```
Comment (adiprerepa) [inline pkg/config/schema/metadata.yaml:259]: why this change?

File: pilot/pkg/config/kube/crdclient/client.go [hunk 3]
```diff
@@ -349,6 +368,23 @@ func (cl *Client) objectInRevision(o *config.Config) bool {
...
+	defer cl.kindsMu.RUnlock()
+	ret := make([]*cacheHandler, 0, len(cl.kinds))
+	for _, k := range cl.kinds {
+		ret = append(ret, k)
+	}
+	return ret
+}
+
+func (cl *Client) kind(r config.GroupVersionKind) (*cacheHandler, bool) {
+	cl.kindsMu.RLock()
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/client.go:382]: Inside handleCRDAdd, which can now happen after startup at any time

File: pkg/config/schema/metadata.yaml (hunk 2 above)
Comment (howardjohn) [reply pkg/config/schema/metadata.yaml:259]: The old way was wrong, but in a way that didn't matter because we didn't use it anywhere. Expected `plural` is all lower case, which is used by all the others.

For example:
```
$ kubectl get crd -oyaml | rg plural:
      plural: authorizationpolicies
      plural: backendconfigs
....
```

**Comment by howardjohn** [comment]
ping @hzxuzhonghu if you have a chance

**Comment by hzxuzhonghu** [comment]
MIssed it

File: pilot/pkg/config/kube/crdclient/cache_handler.go [hunk 4]
```diff
@@ -64,13 +63,14 @@ func (h *cacheHandler) onEvent(old interface{}, curr interface{}, event model.Ev
...
 	// TODO we may consider passing a pointer to handlers instead of the value. While spec is a pointer, the meta will be copied
-	for _, f := range h.handlers {
+	for _, f := range h.client.handlers[h.schema.Resource().GroupVersionKind()] {
 		f(oldConfig, currConfig, event)
 	}
 	return nil
 }
 
 func createCacheHandler(cl *Client, schema collection.Schema, i informers.GenericInformer) *cacheHandler {
+	scope.Debugf("registered CRD %v", schema.Resource())
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/cache_handler.go:73]: schema.Resource().GroupVersionKind()

File: pilot/pkg/config/kube/crdclient/client.go [hunk 5]
```diff
@@ -423,3 +459,56 @@ func genPatchBytes(oldRes, modRes runtime.Object, patchType types.PatchType) ([]
...
+	if !f {
+		scope.Debugf("added resource that we are not watching: %v", name)
+		return
+	}
+	resourceGVK := s.Resource().GroupVersionKind()
+	gvr := s.Resource().GroupVersionResource()
+
+	cl.kindsMu.Lock()
+	defer cl.kindsMu.Unlock()
+	if ch, f := cl.kinds[s.Resource().GroupVersionKind()]; f {
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client.go:475]: cl.kinds[resourceGVK]

File: pilot/pkg/config/kube/crdclient/client.go [hunk 6]
```diff
@@ -187,14 +190,29 @@ func (cl *Client) Run(stop <-chan struct{}) {
...
+
+	cl.crdMetadataInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			crd, ok := obj.(*metav1.PartialObjectMetadata)
+			if !ok {
+				// Shouldn't happen
+				scope.Errorf("wrong type %T: %v", obj, obj)
+				return
+			}
+			handleCRDAdd(cl, crd.Name, stop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client.go:202]: This is time costing, it is not a good practice to synchronously handle event

Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/client.go:202]: I think the only place this may block is in the waitforcachesync part. I may be able to remove that - if so there is nothing doing any IO here so it should be fast enough?

if not maybe can move this into go routine?

File: pilot/pkg/config/kube/crdclient/client.go [hunk 7]
```diff
@@ -423,3 +459,56 @@ func genPatchBytes(oldRes, modRes runtime.Object, patchType types.PatchType) ([]
...
+		// Shouldn't happen
+		scope.Errorf("failed to create informer for %v", resourceGVK)
+		return
+	}
+	cl.kinds[s.Resource().GroupVersionKind()] = createCacheHandler(cl, s, i)
+	if stop != nil {
+		// Start informer factory, only if stop is defined. In startup case, we will not start here as
+		// we will start all factories once we are ready to initialize.
+		// For dynamically added CRDs, we need to start immediately though
+		ifactory.Start(stop)
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client.go:508]: looks there can start multi informer for a single crd, if it is created before istiod start

Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/client.go:508]: the Start code does nothing if called multiple times , so it shouldn't be an issue. It's really "start all unstarted informers"

**Comment by hzxuzhonghu** [comment]
The most critical part is if the crd does not exist, istiod still start the informer, and it list-watch that cr, but with error like this 
```
2495:2021-07-21T03:09:21.565139Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
2496:2021-07-21T03:09:27.141529Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
2497:2021-07-21T03:09:36.565058Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
2498:2021-07-21T03:09:49.502129Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
2499:2021-07-21T03:10:40.019219Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
2500:2021-07-21T03:11:15.959366Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)

```

Then after WorkloadEntry crd created, no error thrown?

**Comment by howardjohn** [comment]
> The most critical part is if the crd does not exist, istiod still start the informer, and it list-watch that cr, but with error like this
> 
> ```
> 2495:2021-07-21T03:09:21.565139Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
> 2496:2021-07-21T03:09:27.141529Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
> 2497:2021-07-21T03:09:36.565058Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
> 2498:2021-07-21T03:09:49.502129Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
> 2499:2021-07-21T03:10:40.019219Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
> 2500:2021-07-21T03:11:15.959366Z	error	watch error in cluster Kubernetes: failed to list *v1alpha3.WorkloadEntry: the server could not find the requested resource (get workloadentries.networking.istio.io)
> ```
> 
> Then after WorkloadEntry crd created, no error thrown?

Is this from my PR? The expectation is:
* at startup,start all informers for know CRDs. This means if one is missing we do not start
* If CRD is later created start it

So I wouldn't expect that to happen (but could be a bug)

**Comment by hzxuzhonghu** [comment]
This is from the master branch

**Comment by hzxuzhonghu** [comment]
>This means if one is missing we do not start

How do you do this?

**Comment by howardjohn** [comment]
> > This means if one is missing we do not start
> 
> How do you do this?
The code for this already exists,unchanged here


> This is from the master branch

That should not happen,let me investigate. We do have a unit test of this but no integ test

**Comment by hzxuzhonghu** [comment]
Yes, that's weird

**Comment by hzxuzhonghu** [comment]
My test way was not right, i deleted the crd after istiod started

**Comment by howardjohn** [comment]
Ah makes sense. Yeah we don't handle deletion well. I think it's a bit harder and not that common of use case

**Comment by howardjohn** [comment]
@hzxuzhonghu I think I have addressed the comments

**Review by hzxuzhonghu — APPROVED** [review]
